File: backend/app/workflow.py
from datetime import datetime, timezone

from fastapi import HTTPException

from .models import Role, StageAction, StageKey, StageStatus


DEPENDENCIES: dict[StageKey, list[StageKey]] = {
    StageKey.MATERIAL: [StageKey.ORDER], StageKey.DESIGN: [StageKey.ORDER],
    StageKey.CUTTING: [StageKey.ORDER], StageKey.PLATE: [StageKey.ORDER],
    StageKey.PRINTING: [StageKey.MATERIAL, StageKey.CUTTING, StageKey.DESIGN, StageKey.PLATE], StageKey.STITCHING: [StageKey.PRINTING],
    StageKey.PACKING: [StageKey.STITCHING], StageKey.DC: [StageKey.PACKING],
    StageKey.BILLING: [StageKey.DC], StageKey.PAYMENT: [StageKey.BILLING],
    StageKey.DISPATCH: [StageKey.BILLING], StageKey.DELIVERY: [StageKey.PAYMENT, StageKey.DISPATCH],
    StageKey.RETURN: [StageKey.DELIVERY], StageKey.REFUND: [StageKey.RETURN],
}

STAGE_ROLES: dict[StageKey, Role] = {
    StageKey.ORDER: Role.ADMIN, StageKey.MATERIAL: Role.CUTTING_MASTER,
    StageKey.DESIGN: Role.DESIGNER, StageKey.CUTTING: Role.CUTTING_MASTER,
    StageKey.PLATE: Role.TRANSPORT_MANAGER, StageKey.PRINTING: Role.PRINTING_OPERATOR,
    StageKey.STITCHING: Role.MANAGER, StageKey.PACKING: Role.MANAGER,
    StageKey.DC: Role.MANAGER, StageKey.BILLING: Role.ACCOUNTANT,
    StageKey.PAYMENT: Role.ACCOUNTANT, StageKey.DISPATCH: Role.TRANSPORT_MANAGER,
    StageKey.DELIVERY: Role.MARKETING, StageKey.RETURN: Role.MARKETING,
    StageKey.REFUND: Role.ACCOUNTANT,
}

LEGACY_ROLE_ALIASES = {
    Role.INVENTORY_MANAGER: Role.CUTTING_MASTER,
    Role.CUTTING_MANAGER: Role.CUTTING_MASTER,
    Role.PLATE_OPERATOR: Role.TRANSPORT_MANAGER,
    Role.DISPATCH_MANAGER: Role.TRANSPORT_MANAGER,
    Role.STITCHING_MANAGER: Role.MANAGER,
    Role.PACKING_MANAGER: Role.MANAGER,
}

SEQUENCE = [StageKey.MATERIAL, StageKey.DESIGN, StageKey.CUTTING, StageKey.PLATE, StageKey.PRINTING,
            StageKey.STITCHING, StageKey.PACKING, StageKey.DC, StageKey.BILLING, StageKey.PAYMENT,
            StageKey.DISPATCH, StageKey.DELIVERY, StageKey.RETURN, StageKey.REFUND]


def initial_stages() -> dict[str, dict]:
    result = {key.value: {"status": StageStatus.WAITING.value, "ownerRole": STAGE_ROLES[key].value} for key in StageKey}
    result[StageKey.ORDER.value] = {"status": StageStatus.COMPLETED.value, "ownerRole": Role.ADMIN.value,
                                    "completedAt": datetime.now(timezone.utc).isoformat()}
    result[StageKey.MATERIAL.value]["status"] = StageStatus.READY.value
    result[StageKey.DESIGN.value]["status"] = StageStatus.READY.value
    result[StageKey.CUTTING.value]["status"] = StageStatus.READY.value
    result[StageKey.PLATE.value]["status"] = StageStatus.READY.value
    return result


def assert_permission(role: Role, stage: StageKey) -> None:
    owner = STAGE_ROLES[stage]
    effective_role = LEGACY_ROLE_ALIASES.get(role, role)
    if owner != effective_role:
        raise HTTPException(
            403,
            f"This step belongs to the {owner.value.replace('_', ' ')}. Ask that team member to update it.",
        )


def ready(order: dict, stage: StageKey) -> bool:
    return all(order["stages"][item.value]["status"] == StageStatus.COMPLETED.value for item in DEPENDENCIES.get(stage, []))


def refresh_ready_states(order: dict) -> None:
    for stage in SEQUENCE:
        state = order["stages"][stage.value]
        if state["status"] == StageStatus.WAITING.value and ready(order, stage):
            state["status"] = StageStatus.READY.value
    candidates = [stage for stage in SEQUENCE if order["stages"][stage.value]["status"] in {
        StageStatus.BLOCKED.value, StageStatus.ISSUE.value, StageStatus.IN_PROGRESS.value, StageStatus.READY.value,
    }]
    order["currentStage"] = (candidates[0] if candidates else StageKey.DELIVERY).value


def apply_action(order: dict, stage: StageKey, action: StageAction, quantity: int | None, note: str | None, data: dict) -> str:
    state = order["stages"][stage.value]
    current = StageStatus(state["status"])
    now = datetime.now(timezone.utc).isoformat()
    if stage == StageKey.DESIGN and action == StageAction.COMPLETE:
        raise HTTPException(409, "Design must be approved or marked as having no customer image.")
    if action == StageAction.START:
        if current != StageStatus.READY or not ready(order, stage):
            raise HTTPException(409, "Previous work is not complete. This step cannot start yet.")
        state.update(status=StageStatus.IN_PROGRESS.value, startedAt=now)
        message = f"Started {stage.value}."
    elif action == StageAction.PROGRESS:
        if current != StageStatus.IN_PROGRESS:
            raise HTTPException(409, "Start this work before updating progress.")
        if quantity is None or quantity > order["quantity"]:
            raise HTTPException(422, "Completed quantity must not exceed the order quantity.")
        state.update(completedQuantity=quantity, progress=round(quantity / order["quantity"] * 100), data=data)
        message = f"Updated {stage.value} progress to {quantity} of {order['quantity']}."
    elif action == StageAction.BLOCK:
        if not note:
            raise HTTPException(422, "Explain the issue so the next person knows what to resolve.")
        state.update(status=StageStatus.BLOCKED.value, note=note)
        message = f"Blocked {stage.value}: {note}"
    elif action == StageAction.RESOLVE:
        if current not in {StageStatus.BLOCKED, StageStatus.ISSUE}:
            raise HTTPException(409, "This step has no open issue.")
        state.update(status=StageStatus.READY.value, note=note or "Issue resolved.")
        message = f"Resolved the {stage.value} issue."
    else:
        if current != StageStatus.IN_PROGRESS:
            raise HTTPException(409, "Start this work before marking it complete.")
        state.update(status=StageStatus.COMPLETED.value, completedAt=now, progress=100,
                     completedQuantity=order["quantity"], data=data, note=note)
        message = f"Completed {stage.value}."
    refresh_ready_states(order)
    return message

File: src/lib/workflow.ts
import type { Order, Role, StageKey, StageStatus } from '../types'

export const stageInfo: Record<StageKey, { label: string; short: string; role: Role; help: string }> = {
  order: { label: 'Order booked', short: 'Order', role: 'admin', help: 'Customer order is confirmed and ready for preparation.' },
  material: { label: 'Material available', short: 'Material', role: 'cutting_master', help: 'Check and reserve the required bag material.' },
  design: { label: 'Design approved', short: 'Design', role: 'designer', help: 'Prepare the artwork and receive approval.' },
  cutting: { label: 'Cutting', short: 'Cutting', role: 'cutting_master', help: 'Cut material to the required bag specification.' },
  plate: { label: 'Plate preparation', short: 'Plate', role: 'transport_manager', help: 'Prepare the printing plate from the approved design.' },
  printing: { label: 'Printing', short: 'Printing', role: 'printing_operator', help: 'Printing starts after cutting and plate preparation are complete.' },
  stitching: { label: 'Stitching', short: 'Stitching', role: 'manager', help: 'Stitch the printed pieces and update completed quantity.' },
  packing: { label: 'Packing', short: 'Packing', role: 'manager', help: 'Count, pack, weigh, and prepare boxes.' },
  dc: { label: 'Delivery challan', short: 'D.C.', role: 'manager', help: 'Generate the delivery challan for the packed order.' },
  billing: { label: 'Billing', short: 'Billing', role: 'accountant', help: 'Create and verify the customer invoice.' },
  payment: { label: 'Payment', short: 'Payment', role: 'accountant', help: 'Record payment received from the customer.' },
  dispatch: { label: 'Dispatch', short: 'Dispatch', role: 'transport_manager', help: 'Record transporter and tracking information.' },
  delivery: { label: 'Delivery confirmation', short: 'Delivery', role: 'marketing', help: 'Confirm that the customer received the order.' },
  return: { label: 'Return', short: 'Return', role: 'marketing', help: 'Record returned goods only when applicable.' },
  refund: { label: 'Refund', short: 'Refund', role: 'accountant', help: 'Record an approved refund only when applicable.' },
}

export const dependencies: Partial<Record<StageKey, StageKey[]>> = {
  material: ['order'], design: ['order'], cutting: ['order'], plate: ['order'],
  printing: ['material', 'cutting', 'design', 'plate'], stitching: ['printing'], packing: ['stitching'],
  dc: ['packing'], billing: ['dc'], payment: ['billing'], dispatch: ['billing'],
  delivery: ['payment', 'dispatch'], return: ['delivery'], refund: ['return'],
}

export const productionStages: StageKey[] = ['material', 'design', 'cutting', 'plate', 'printing', 'stitching', 'packing', 'dc', 'billing', 'payment', 'dispatch', 'delivery']

export function isReady(order: Order, stage: StageKey) {
  return (dependencies[stage] ?? []).every((key) => order.stages[key].status === 'completed')
}

export function operatingRole(role: Role): Role {
  return ({ inventory_manager: 'cutting_master', cutting_manager: 'cutting_master', plate_operator: 'transport_manager', dispatch_manager: 'transport_manager', stitching_manager: 'manager', packing_manager: 'manager' } as Partial<Record<Role, Role>>)[role] ?? role
}

export function canManageStage(role: Role, stage: StageKey) {
  // Admin can supervise every order, but operational facts must be entered by
  // the department that performed the work (for example, stock by Material).
  return stageInfo[stage].role === operatingRole(role)
}

export function statusLabel(status: StageStatus) {
  return ({
    not_started: 'Not started', waiting: 'Waiting', ready: 'Ready to start', in_progress: 'In progress',
    completed: 'Completed', blocked: 'Blocked', issue: 'Issue reported', not_applicable: 'Not applicable',
  } satisfies Record<StageStatus, string>)[status]
}

export function nextInstruction(order: Order) {
  const stage = order.currentStage
  const state = order.stages[stage]
  if (state.status === 'blocked' || state.status === 'issue') return `Resolve the ${stageInfo[stage].short.toLowerCase()} issue before production can continue.`
  if (state.status === 'in_progress') return `Continue ${stageInfo[stage].short.toLowerCase()} and update the completed quantity.`
  return `The order is ready for ${stageInfo[stage].short.toLowerCase()}. The responsible team can start now.`
}
